src/Network/MQTT/Client.hs

```diff
@@ -0,0 +1,190 @@
+      if null ps
+        then retry
+        else forM_ ps $ writeTQueue receiverQueue
+
+main :: IO ()
+main = do
+  socketHandle <- conn "test.mosca.io" "1883"
+
+  clientState <- atomically initState
+
+  forkIO $ internalSender clientState socketHandle
+  forkIO $ internalReceiver clientState socketHandle
+
+  forkIO $ receiver clientState
+  forkIO $ do
+    sender clientState connectPacket
+--    sender clientState pingReq
+    threadDelay sec
+    sender clientState publishPacketQoS1
+
+  threadDelay $ 3 * sec
+
+------------------------------------------------------------------------------------------------
+
+sec = 1000000
+
+data ClientState =
+  ClientState
+  { senderQueue    :: TQueue Packet
+  , receiverQueue  :: TQueue Packet
+  , qosLevel       :: TVar (Maybe QoS)
+  , sessionPresent :: TVar (Maybe Bool)
+  , pingResp       :: TVar (Maybe Bool)
+  }
+  deriving (Eq)
+
+initState :: STM ClientState
+initState = do
+  senderQueue <- newTQueue
+  receiverQueue <- newTQueue
+  qosLevel <- newTVar Nothing
+  isSessionPresent <- newTVar Nothing
+  pingResp <- newTVar Nothing
+  return $ ClientState
+             senderQueue
+             receiverQueue
+             qosLevel
+             isSessionPresent
+             pingResp
+
+------------------------------------------------------------------------------------------------
+
+pingReq = PINGREQ PingreqPacket
+
+publishPacketQoS0 =
+  PUBLISH PublishPacket { publishDup = False
+                        , publishMessage = Message QoS0 False (Topic $ T.pack "a/b") BS.empty
+                        , publishPacketIdentifier = Nothing
+                        }
+
+publishPacketQoS1 =
+  PUBLISH PublishPacket { publishDup = True
+                        , publishMessage = Message QoS1 True (Topic $ T.pack "a/b") (BS.pack [ 0xab, 0x12 ])
+                        , publishPacketIdentifier = Just (PacketIdentifier 10)
+                        }
+
+
+publishPacketQoS2 =
+  PUBLISH PublishPacket { publishDup = True
+                        , publishMessage = Message QoS2 True (Topic $ T.pack "a/b") (BS.pack [ 0xab, 0x12 ])
+                        , publishPacketIdentifier = Just (PacketIdentifier 10)
+                        }
+
+connectPacket =
+  CONNECT ConnectPacket { connectClientIdentifier = ClientIdentifier (T.pack "abc")
+                        , connectProtocolLevel = 0x04
+                        , connectWillMsg = Just $
+                            Message QoS1
+                                    False
+                                    (Topic $ T.pack "a/b")
+                                    (BS.pack [ 0x12, 0xab ])
+                        , connectUserName = Just (UserName $ T.pack "ac")
+                        , connectPassword = Just (Password $
+                                                      BS.pack [ 0xcd, 0xbb, 0x11 ])
+                        , connectCleanSession = True
+                        , connectKeepAlive = 0x000c
```

Comment:
You don't send pingreq periodically to keep session active, so keepalive should be set to 0 (disabled).